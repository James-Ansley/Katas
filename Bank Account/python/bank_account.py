from collections.abc import Iterable, Iterator
from dataclasses import dataclass
from datetime import datetime
from typing import Final, Self


class BankAccount:
    def __init__(self):
        self._transactions: Final[list[Transaction]] = []
        self._table = Table(
            separator=" || ",
            header=("Date", "Amount", "Balance"),
            formats=("{:%Y-%m-%d}", "{:<8}", "{:<8}"),
            header_formats=("{:<10}", "{:<8}", "{:<8}"),
        )

    def deposit(self, amount: int) -> None:
        transaction = Transaction.new_deposit(amount)
        self._transactions.append(transaction)

    def withdraw(self, amount: int) -> None:
        transaction = Transaction.new_withdrawal(amount)
        self._transactions.append(transaction)

    def print_statement(self) -> None:
        print(self._table.header())
        for date, amount, balance in self._transaction_history():
            row = self._table.row(date, amount, balance)
            print(row)

    def _transaction_history(self) -> Iterator[tuple[datetime, int, int]]:
        result = []
        running_total = 0
        for transaction in sorted(self._transactions, key=lambda t: t.date):
            running_total += transaction.amount
            trans_info = (transaction.date, transaction.amount, running_total)
            result.append(trans_info)
        return reversed(result)


@dataclass
class Transaction:
    date: datetime
    amount: int

    @classmethod
    def new_deposit(cls, amount) -> Self:
        if amount <= 0:
            raise ValueError("Transactions cannot be non-positive")
        return Transaction(datetime.now(), amount)

    @classmethod
    def new_withdrawal(cls, amount) -> Self:
        if amount <= 0:
            raise ValueError("Transactions cannot be non-positive")
        return Transaction(datetime.now(), -amount)


class Table:
    def __init__(
          self,
          *,
          separator: str,
          header: Iterable[str],
          formats: Iterable[str],
          header_formats: Iterable[str] = None,
    ):
        self.separator = separator
        self.column_headers = header
        self.formats = [fmt.format for fmt in formats]

        if header_formats is not None:
            self.header_formats = [fmt.format for fmt in header_formats]
        else:
            self.header_formats = self.formats

    def header(self):
        zipped_headers_and_fmt = zip(self.column_headers, self.header_formats)
        return self.separator.join(
            fmt(cell) for cell, fmt in zipped_headers_and_fmt
        )

    def row(self, *values):
        return self.separator.join(
            fmt(cell) for cell, fmt in zip(values, self.formats)
        )
