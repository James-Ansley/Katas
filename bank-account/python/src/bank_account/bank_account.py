from datetime import datetime


class BankAccount:
    def __init__(self):
        self._transactions = {}

    def deposit(self, amount: int) -> None:
        today = datetime.today()
        self._transactions[today] = amount

    def withdraw(self, amount: int) -> None:
        today = datetime.today()
        self._transactions[today] = -amount

    def print_statement(self) -> None:
        balance = 0
        result = []
        for date, amount in sorted(self._transactions.items()):
            balance += amount
            result.append(_format_transaction(date, amount, balance))
        result.append("Date       || Amount || Balance")
        print("\n".join(reversed(result)))


def _format_transaction(date, amount, balance):
    return " || ".join((
        f"{date.strftime('%Y-%m-%d'):>10}",
        f"{amount:<6}",
        f"{balance:<7}",
    ))
