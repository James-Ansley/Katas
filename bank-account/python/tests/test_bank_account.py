from contextlib import redirect_stdout
from io import StringIO

from approvaltests import verify
from freezegun import freeze_time

from bank_account import BankAccount
from utils import test


@test
def a_new_bank_account_will_display_a_blank_statement():
    bank_account = BankAccount()
    with redirect_stdout(StringIO()) as f:
        bank_account.print_statement()
    verify(f.getvalue())


@test
def a_bank_account_will_display_a_single_deposit_in_its_statement():
    bank_account = BankAccount()

    with freeze_time("2012-01-10"):
        bank_account.deposit(123)

    with redirect_stdout(StringIO()) as f:
        bank_account.print_statement()
    verify(f.getvalue())


@test
def a_bank_account_will_display_a_single_withdrawal_in_its_statement():
    bank_account = BankAccount()

    with freeze_time("2012-01-14"):
        bank_account.withdraw(123)

    with redirect_stdout(StringIO()) as f:
        bank_account.print_statement()
    verify(f.getvalue())


@test
def a_bank_account_will_display_multiple_deposits_in_its_statement():
    bank_account = BankAccount()

    with freeze_time("2012-01-10"):
        bank_account.deposit(1000)
    with freeze_time("2012-01-13"):
        bank_account.deposit(2000)

    with redirect_stdout(StringIO()) as f:
        bank_account.print_statement()
    verify(f.getvalue())


@test
def a_bank_account_will_display_multiple_withdrawals_in_its_statement():
    bank_account = BankAccount()

    with freeze_time("2012-01-10"):
        bank_account.withdraw(1000)
    with freeze_time("2012-01-13"):
        bank_account.withdraw(2000)

    with redirect_stdout(StringIO()) as f:
        bank_account.print_statement()
    verify(f.getvalue())


@test
def a_bank_account_will_display_multiple_transactions_in_its_statement():
    bank_account = BankAccount()

    with freeze_time("2012-01-10"):
        bank_account.deposit(1000)
    with freeze_time("2012-01-13"):
        bank_account.deposit(2000)
    with freeze_time("2012-01-14"):
        bank_account.withdraw(500)

    with redirect_stdout(StringIO()) as f:
        bank_account.print_statement()
    verify(f.getvalue())
