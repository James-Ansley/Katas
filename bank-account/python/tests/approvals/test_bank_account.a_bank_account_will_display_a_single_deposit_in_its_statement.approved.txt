Date       || Amount || Balance
2012-01-10 || 123    || 123    
